File: src/main/java/org/extrasapache/myfaces/codi/examples/ebean/support/data/PaginationController.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package org.extrasapache.myfaces.codi.examples.ebean.support.data;

import com.avaje.ebean.Page;
import com.avaje.ebean.PagingList;

import java.util.*;
import java.util.concurrent.Future;

/**
 * @author Werner Punz (latest modification by $Author$)
 * @version $Revision$ $Date$
 */

public class PaginationController<T> {
    PagingList<T> _delegate;
    Integer lastPageAccessed = 0;

    int pagingWindowSize = 8;

    public PaginationController(PagingList<T> delegate) {

        _delegate = delegate;
    }

    public void refresh() {
        _delegate.refresh();
    }

    public PagingList<T> setFetchAhead(boolean b) {
        return _delegate.setFetchAhead(b);
    }

    public Future<Integer> getFutureRowCount() {
        return _delegate.getFutureRowCount();
    }

    public List<T> getAsList() {
        return _delegate.getAsList();
    }

    public int getPageSize() {
        return _delegate.getPageSize();
    }

    public int getTotalRowCount() {
        return _delegate.getTotalRowCount();
    }

    public int getTotalPageCount() {
        return _delegate.getTotalPageCount();
    }

    public Page<T> getPage(int i) {
        lastPageAccessed = i;
        return _delegate.getPage(i);
    }

    public Page<T> fetchPage() {
        return _delegate.getPage(lastPageAccessed);
    }

    public Integer getLastPageAccessed() {
        return lastPageAccessed;
    }

    public void setLastPageAccessed(Integer lastPageAccessed) {
        this.lastPageAccessed = lastPageAccessed;
    }

    public boolean isFirstPage() {
        return this.lastPageAccessed == 0;
    }

    public boolean isLastPage() {
        return this.lastPageAccessed >= this.getTotalPageCount() - 1;
    }

    public boolean isInFirstWindow() {
        if (getPageSize() == 0) {
            return true;
        }
        return lastPageAccessed < pagingWindowSize;
    }

    public boolean hasPreviousPagesWindow() {
        return false;
    }

    public Integer getFirstPageIdx() {

        return 0;
    }

    public Integer getLastPageIdx() {

        return _delegate.getTotalPageCount() - 1;
    }

    public Integer getPreviousPage() {
        return Math.max(0, this.lastPageAccessed -1);
    }

    public Integer getNextPage() {
        return Math.max(0,Math.min(this.lastPageAccessed, this.getTotalPageCount() -1));
    }

    public Integer getNextWindowIdx() {
        if (isInLastWindow()) return null;
        return (int) (Math.floor((double) lastPageAccessed / (double) pagingWindowSize) + 1) * pagingWindowSize;
    }

    public Integer getPreviousWindowIdx() {
        if (isInFirstWindow()) return null;
        return (int) (Math.floor((double) lastPageAccessed / (double) pagingWindowSize) - 1) * pagingWindowSize;
    }

    public boolean isInLastWindow() {
        if (getPageSize() == 0) {
            return true;
        }
        return lastPageAccessed >= getTotalPageCount() - pagingWindowSize;
    }

    public List<Integer> fetchPagesWindow() {
        List<Integer> ret = new ArrayList<Integer>(getPageSize());

        if (getTotalPageCount() == 1) {
            return ret;
        }

        int pagingWindowPos = (int) Math.floor((double) lastPageAccessed / (double) pagingWindowSize);

        for (int cnt = pagingWindowPos; cnt < getTotalPageCount(); cnt++) {
            ret.add(cnt);
        }
        return ret;
    }

    public List getPageAsList() {
        return fetchPage().getList();
    }
}

File: src/main/scala/org/extrasapache/myfaces/codi/examples/ebean/support/data/PaginationController.scala
package org.extrasapache.myfaces.codi.examples.ebean.support.data

import reflect.BeanProperty
import java.util.concurrent.Future
import com.avaje.ebean.{Page, PagingList}

import scala.math._
import java.util._

/**
 *
 * @author Werner Punz (latest modification by $Author$)
 * @version $Revision$ $Date$
 *
 * This is our pagination controller, not most of the times
 * we can get away by using the Int -> int(java) autoboxing
 * but since scala does not know any Integer -> int autoboxing
 * we cannot get away the other way
 *
 * Also a null value enforces integer to be used
 */

class PaginationController[T](var _delegate: PagingList[T]) {
  //var _delegate: PagingList[T] = _
  @BeanProperty
  var lastPageAccessed: Int = 0
  var pagingWindowSize: Int = 8

  def refresh() = _delegate.refresh

  def setFetchAhead(b: Boolean) = _delegate.setFetchAhead(b)

  def getFutureRowCount: Future[java.lang.Integer] = _delegate.getFutureRowCount

  def getAsList: java.util.List[T] = _delegate.getAsList

  def getPageSize: Int = _delegate.getPageSize

  def getTotalRowCount: Int = _delegate.getTotalRowCount

  def getTotalPageCount: Int = _delegate.getTotalPageCount

  def getPage(i: Int): Page[T] = {
    lastPageAccessed = i
    _delegate.getPage(i)
  }

  def fetchPage: Page[T] = {
    _delegate.getPage(lastPageAccessed)
  }

  def isFirstPage: Boolean = lastPageAccessed == 0

  def isLastPage: Boolean = lastPageAccessed >= getTotalPageCount

  def isInFirstWindow: Boolean = (getPageSize == 0) || (lastPageAccessed < pagingWindowSize)

  def hasPreviousPagesWindow: Boolean = false

  def getFirstPageIdx: Int = 0

  def getLastPageIdx: Int = _delegate.getTotalPageCount - 1

  def getPreviousPage: Int = {
    val ret = max(0, lastPageAccessed - 1)
    ret
  }



  def getNextWindowIdx: java.lang.Integer = {
    if (isInLastWindow)
      null
    else
      (floor(lastPageAccessed.doubleValue / pagingWindowSize.doubleValue).intValue + 1) * pagingWindowSize
  }

  def getPreviousWindowIdx: java.lang.Integer = {
    if (isInFirstWindow) null
    else {
      (floor(lastPageAccessed.doubleValue / pagingWindowSize.doubleValue).intValue - 1) * pagingWindowSize
    }
  }

  def isInLastWindow: Boolean = {
    (getPageSize == 0) || (lastPageAccessed >= getTotalPageCount - pagingWindowSize)
  }

  def fetchPagesWindow: List[Int] = {
    val ret = new ArrayList[Int](getPageSize)
    if (getTotalPageCount != 1) {
      val pagingWindowPos = floor(lastPageAccessed.doubleValue / pagingWindowSize.doubleValue).intValue
      (pagingWindowPos until getTotalPageCount).foreach {
        ret.add(_)
      }
    }
    ret
  }

  def getPageAsList: List[T] = {
    fetchPage.getList
  }

  def getNextPage: Int = {
    max(0, min(lastPageAccessed, getTotalPageCount - 1))
  }

}